docs/architecture/adr-008-chain-endpoint-handle-consolidation.md

```diff
@@ -0,0 +1,123 @@
+```rust
+// In ChainHandle
+fn query_host_consensus_state(&self, height: Height) -> Result<AnyConsensusState, Error>;
+
+// In ChainEndpoint
+fn query_host_consensus_state(&self, height: ICSHeight) -> Result<Self::ConsensusState, Error>;
+```
+
+These will end up as
+```rust
+fn query_host_consensus_state(&self, height: ICSHeight) -> Result<Self::ConsensusState, Error>;
+```
+in `ChainBase`; that is, they will return an associated type. Then, `ChainHandle` would be declared as
+```rust
+trait ChainHandle: ChainBase<ConsensusState=AnyConsensusState, ...> { ... }
+```
+
+
+### Procedural macro
+We propose a procedural macro that will enhance readability of `query_*` methods. For example, in `BaseChainHandle`, the pattern looks like:
+```rust
+fn query_client_connections(
+    &self,
+    request: QueryClientConnectionsRequest,
+) -> Result<Vec<ConnectionId>, Error> {
+    self.send(|reply_to| ChainRequest::QueryClientConnections { request, reply_to })
+}
+
+fn query_consensus_states(
+    &self,
+    request: QueryConsensusStatesRequest,
+) -> Result<Vec<AnyConsensusStateWithHeight>, Error> {
+    self.send(|reply_to| ChainRequest::QueryConsensusStates { request, reply_to })
+}
+
+...
+```
+
+The highly repetitive implementation pattern makes it tedious to maintain, as well as see the pattern itself: a developer should check all implementations to make sure that there isn't a method which is implemented slightly differently than the others. We suggest solving this with a procedural macro.
+
+```rust
+#[unroll_template(
+    [query_client_connections, QueryClientConnectionsRequest, ChainRequest::QueryClientConnections],
+    [query_consensus_states, QueryClientConsensusStatesRequest, ChainRequest::QueryConsensusStates],
+)]
+fn _0(&self, request: _1) {
+    self.send(|reply_to| _2 { request, reply_to })
+}
+```
+
+In the macro expansion phase, the function definition will be duplicated once per argument provided to `unroll_template()`. Here, notice that 2 arguments are provided to `unroll_template`, 2 lists, and therefore the function definition would be duplicated twice. In each instance of the duplicated function, the markers "_0", "_1", and "_2" will be replaced by the expression at index `_i`. For example, in the first function, `_0` is replaced by `query_client_connections`, `_1` by `QueryClientConnectionsRequest`, and `_2` by `ChainRequest::QueryClientConnections`. So, the expanded code would be:
+
+```rust
+fn query_client_connections(
+    &self,
+    request: QueryClientConnectionsRequest,
+) -> Result<Vec<ConnectionId>, Error> {
+    self.send(|reply_to| ChainRequest::QueryClientConnections { request, reply_to })
+}
+
+fn query_consensus_states(
+    &self,
+    request: QueryConsensusStatesRequest,
+) -> Result<Vec<AnyConsensusStateWithHeight>, Error> {
+    self.send(|reply_to| ChainRequest::QueryConsensusStates { request, reply_to })
+}
+```
+
+## Status
+
+Proposed
```

Comment:
Let's also record the status in the table here https://github.com/informalsystems/ibc-rs/tree/master/docs/architecture#table-of-contents